File: ios_app/src/services/OrangeClient.ts
import { ConsumptionData, ConsumptionDetail } from '../types';

export class OrangeClient {
  private userAgent =
    'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/150.0.0.0 Safari/537.36';
  private activeMsisdn: string;

  constructor(
    private loginId: string,
    private pass: string,
    private targetPhone: string = loginId,
    private savedCookies: string[] = []
  ) {
    this.activeMsisdn = targetPhone;
  }

  public get currentCookies(): string[] {
    return this.savedCookies;
  }

  public async fetchConsumption(): Promise<ConsumptionData> {
    // 1. Try to load /mon-solde directly with stored session
    try {
      const fastRes = await fetch('https://espace-client.orange.ma/mon-solde', {
        headers: {
          'User-Agent': this.userAgent,
          'Accept':
            'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8',
        },
      });
      const html = await fastRes.text();
      if (html.includes('Solde de recharge') || html.includes('mbm')) {
        return this.parseDashboard(html);
      }
    } catch (_e) {}

    // 2. Perform Full 3-Step Login Flow
    return await this.attemptFullLogin();
  }

  /**
   * Extracts the full outer HTML of a balanced <div> starting at startIndex
   */
  private extractBalancedDiv(html: string, startIndex: number): string {
    const tagRegex = /<div\b[^>]*>|<\/div>/gi;
    tagRegex.lastIndex = startIndex;
    let depth = 0;
    let match: RegExpExecArray | null;

    while ((match = tagRegex.exec(html)) !== null) {
      if (match[0].toLowerCase().startsWith('<div')) {
        depth++;
      } else {
        depth--;
        if (depth === 0) {
          return html.slice(startIndex, match.index + match[0].length);
        }
      }
    }
    return '';
  }

  private parseDashboard(html: string): ConsumptionData {
    // 1. Find all div.mbm containers using balanced tag matching (matching Jsoup in Android)
    const mbmOpenings = /<div\b[^>]*class=["\'][^"\']*\bmbm\b[^"\']*["\'][^>]*>/gi;
    let targetSection = '';
    let match: RegExpExecArray | null;

    while ((match = mbmOpenings.exec(html)) !== null) {
      const block = this.extractBalancedDiv(html, match.index);
      if (block.includes('Solde de recharge')) {
        targetSection = block;
        break;
      }
    }

    // 2. Fallback: If not found via class mbm, locate "Solde de recharge" and its enclosing container
    if (!targetSection && html.includes('Solde de recharge')) {
      const soldePos = html.indexOf('Solde de recharge');
      const lastDiv = html.lastIndexOf('<div', soldePos);
      if (lastDiv !== -1) {
        targetSection = this.extractBalancedDiv(html, lastDiv);
      }
      if (!targetSection) {
        // Safe window bounded strictly to Solde de recharge, avoiding header ads
        targetSection = html.slice(soldePos, soldePos + 2500);
      }
    }

    // Strip HTML tags to get pure text strictly from targetSection
    const cleanText = targetSection
      ? targetSection
          .replace(/<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>/gi, '')
          .replace(/<style\b[^<]*(?:(?!<\/style>)<[^<]*)*<\/style>/gi, '')
          .replace(/<[^>]+>/g, ' ')
          .replace(/\s+/g, ' ')
      : '';

    const dhRegex = /\b(\d+(?:[.,]\d+)?)\s*Dh\b/i;
    const internetRegex = /\b(\d+(?:[.,]\d+)?)\s*(MO|GO|MB|GB)\b/i;
    const callsRegex = /\b(\d+h\s*\d+min(?:\s*\d+s)?|\d+min)\b/gi;
    const smsRegex = /\b(\d+)\s*SMS\b/i;

    const dhMatch = cleanText.match(dhRegex);
    const internetMatch = cleanText.match(internetRegex);
    const callsMatches: string[] = [];
    let cMatch: RegExpExecArray | null;
    while ((cMatch = callsRegex.exec(cleanText)) !== null) {
      callsMatches.push(cMatch[1]);
    }
    const smsMatch = cleanText.match(smsRegex);

    const phoneMatch = html.match(/\b(0[67]\d{8})\b/);
    const displayPhone = this.activeMsisdn.includes('@')
      ? (phoneMatch ? phoneMatch[1] : this.activeMsisdn)
      : this.activeMsisdn;

    const soldeDh = dhMatch ? dhMatch[0] : '0.00 Dh';
    const internet = internetMatch ? internetMatch[0] : '0 GO';
    const callsNational = callsMatches[0] || '0h 0min';
    const callsOrange = callsMatches[1] || '0h 0min';
    const callsInter = callsMatches[2] || '0h 0min';
    const sms = smsMatch ? smsMatch[0] : '0 SMS';

    const structuredDetails: ConsumptionDetail[] = [
      { label: 'Solde principal', value: soldeDh, iconType: 'wallet' },
      { label: 'Internet', value: internet, iconType: 'internet' },
      { label: 'Appels Nationaux', value: callsNational, iconType: 'calls' },
      { label: 'Appels Orange', value: callsOrange, iconType: 'orange' },
      { label: 'Appels Internationaux', value: callsInter, iconType: 'global' },
      { label: 'SMS', value: sms, iconType: 'sms' },
    ];

    return {
      operator: 'Orange',
      phoneNumber: displayPhone,
      callsRemaining: callsNational,
      callsPercent: null,
      internetRemaining: internet,
      internetPercent: null,
      extraDetails: soldeDh,
      structuredDetails,
    };
  }

  private async attemptFullLogin(): Promise<ConsumptionData> {
    const loginUrl = 'https://espace-client.orange.ma/sso/login';
    const getRes = await fetch(loginUrl, {
      headers: { 'User-Agent': this.userAgent },
    });
    const tokenHtml = await getRes.text();

    const tokenMatch1 = tokenHtml.match(
      /name=["\']login_form\[ezxform_token\]["\'][^>]*value=["\']([^"\']+)["\']/i
    );
    let token = tokenMatch1 ? tokenMatch1[1] : '';

    // Step 2: POST phone number
    const post1Params = new URLSearchParams();
    post1Params.append('login_form[login]', this.loginId);
    post1Params.append('login_form[ezxform_token]', token);

    const post1Res = await fetch(loginUrl, {
      method: 'POST',
      headers: {
        'User-Agent': this.userAgent,
        'Content-Type': 'application/x-www-form-urlencoded',
        'Origin': 'https://espace-client.orange.ma',
        'Referer': loginUrl,
      },
      body: post1Params.toString(),
    });
    const pwdHtml = await post1Res.text();

    const tokenMatch2 = pwdHtml.match(
      /name=["\']login_form\[ezxform_token\]["\'][^>]*value=["\']([^"\']+)["\']/i
    );
    if (tokenMatch2 && tokenMatch2[1]) {
      token = tokenMatch2[1];
    }

    // Step 3: POST password to /sso/check
    const checkUrl = 'https://espace-client.orange.ma/sso/check';
    const checkParams = new URLSearchParams();
    checkParams.append('login_form[_username]', this.loginId);
    checkParams.append('login_form[_password]', this.pass);
    checkParams.append('login_form[_remember_me]', '1');
    checkParams.append('login_form[ezxform_token]', token);

    const checkRes = await fetch(checkUrl, {
      method: 'POST',
      headers: {
        'User-Agent': this.userAgent,
        'Content-Type': 'application/x-www-form-urlencoded',
        'X-Requested-With': 'XMLHttpRequest',
        'Referer': loginUrl,
      },
      body: checkParams.toString(),
    });
    const checkBody = await checkRes.text();

    if (checkBody.includes('identifiants sont incorrects')) {
      throw new Error('Identifiant ou mot de passe Orange incorrect');
    }

    // Step 4: Fetch /mon-solde
    const soldeRes = await fetch('https://espace-client.orange.ma/mon-solde', {
      headers: {
        'User-Agent': this.userAgent,
        'Referer': 'https://espace-client.orange.ma/',
      },
    });
    const soldeHtml = await soldeRes.text();

    return this.parseDashboard(soldeHtml);
  }
}

File: android_app/app/src/main/java/com/telecom/widget/network/OrangeClient.kt
package com.telecom.widget.network

import com.google.gson.Gson
import com.google.gson.JsonObject
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.FormBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.ResponseBody
import org.jsoup.Jsoup
import org.jsoup.nodes.Element
import java.io.IOException

class OrangeClient(
    private val loginId: String,
    private val pass: String,
    private val targetPhone: String = loginId,
    cookieStrings: List<String> = emptyList()
) : TelecomClient {

    private var activeMsisdn: String = targetPhone

    private val cookieJar = InMemoryCookieJar(parseCookies("https://espace-client.orange.ma", cookieStrings))

    private val client = OkHttpClient.Builder()
        .followRedirects(false)
        .cookieJar(cookieJar)
        .build()

    private val redirectingClient = OkHttpClient.Builder()
        .followRedirects(true)
        .cookieJar(cookieJar)
        .build()

    // Serialize cookies exactly like APK: it.toString()
    override val currentCookies: List<String>
        get() = cookieJar.getAllCookies().map { it.toString() }

    private fun req(url: String, referer: String? = null): Request.Builder {
        val b = Request.Builder()
            .url(url)
            .header("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/150.0.0.0 Safari/537.36")
            .header("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8")
            .header("Accept-Language", "fr-FR,fr;q=0.9,en-US;q=0.8,en;q=0.7,ar;q=0.6")
        if (referer != null) b.header("Referer", referer)
        return b
    }

    // Exactly as in APK: iterate all elements and collect ownText()
    private fun getTextWithNewlines(html: String): String {
        val doc = Jsoup.parse(html)
        val lines = mutableListOf<String>()
        for (el in doc.select("*")) {
            val own = el.ownText().trim()
            if (own.isNotEmpty()) lines.add(own)
        }
        return lines.joinToString("\n")
    }

    private fun executeConsumptionFetch(html: String): ConsumptionData {
        val soup = Jsoup.parse(html)
        // APK selects: div.mbm, then keeps only those containing "Solde de recharge"
        val matchingDivs: List<String> = soup.select("div.mbm").mapNotNull { div: Element ->
            val text = getTextWithNewlines(div.outerHtml())
            if (text.contains("Solde de recharge")) text else null
        }

        if (matchingDivs.isEmpty()) {
            throw IOException("Could not parse Orange consumption. Login may have failed.")
        }

        val fullText = matchingDivs.first()
            .lines().map { it.trim() }.filter { it.isNotBlank() }
            .joinToString("\n")

        val dhRegex = Regex("""\b(\d+(?:\.\d+)?)\s*Dh\b""", RegexOption.IGNORE_CASE)
        val internetRegex = Regex("""\b(\d+(?:\.\d+)?)\s*(MO|GO|MB|GB)\b""", RegexOption.IGNORE_CASE)
        val callsRegex = Regex("""\b(\d+h\s*\d+min(?:\s*\d+s)?)\b""", RegexOption.IGNORE_CASE)
        val smsRegex = Regex("""\b(\d+)\s*SMS\b""", RegexOption.IGNORE_CASE)

        val soldeDh = dhRegex.find(fullText)?.value ?: "0.00 Dh"
        val internet = internetRegex.find(fullText)?.value ?: "0 GO"
        val callsMatches = callsRegex.findAll(fullText).toList()
        val callsNational = callsMatches.getOrNull(0)?.value ?: "0h 0min"
        val callsOrange = callsMatches.getOrNull(1)?.value ?: "0h 0min"
        val callsInter = callsMatches.getOrNull(2)?.value ?: "0h 0min"
        val sms = smsRegex.find(fullText)?.value ?: "0 SMS"

        val phoneMatch = Regex("""\b(0[67]\d{8})\b""").find(html)
        val displayPhone = if (activeMsisdn.contains("@")) phoneMatch?.value ?: activeMsisdn else activeMsisdn

        val structuredDetails = listOf(
            ConsumptionDetail("Solde principal", soldeDh, "wallet"),
            ConsumptionDetail("Internet", internet, "internet"),
            ConsumptionDetail("Appels Nationaux", callsNational, "calls"),
            ConsumptionDetail("Appels Orange", callsOrange, "orange"),
            ConsumptionDetail("Appels Internationaux", callsInter, "global"),
            ConsumptionDetail("SMS", sms, "sms")
        )

        return ConsumptionData(
            operator = "Orange",
            phoneNumber = displayPhone,
            callsRemaining = callsNational,
            internetRemaining = internet,
            extraDetails = soldeDh,
            structuredDetails = structuredDetails
        )
    }

    suspend fun submitLineSelection(line: String, token: String): ConsumptionData = withContext(Dispatchers.IO) {
        activeMsisdn = line
        redirectingClient.newCall(
            req("https://espace-client.orange.ma/select-msisdn", "https://espace-client.orange.ma/sso/login")
                .header("Origin", "https://espace-client.orange.ma")
                .post(FormBody.Builder()
                    .add("select_msisdn_form[msisdn]", line)
                    .add("select_msisdn_form[ezxform_token]", token)
                    .build())
                .build()
        ).execute()

        val body: ResponseBody? = redirectingClient.newCall(
            req("https://espace-client.orange.ma/mon-solde", "https://espace-client.orange.ma/").build()
        ).execute().body
        val html = body?.string() ?: ""
        executeConsumptionFetch(html)
    }

    override suspend fun fetchConsumption(): ConsumptionData = withContext(Dispatchers.IO) {
        // Try with existing cookies first
        if (cookieJar.getAllCookies().isNotEmpty()) {
            try {
                val fastRes = redirectingClient.newCall(
                    req("https://espace-client.orange.ma/mon-solde").build()
                ).execute()
                val urlStr = fastRes.request.url.toString()
                if (fastRes.isSuccessful && !urlStr.contains("sso/login")) {
                    val html = fastRes.body?.string() ?: ""
                    if (html.contains("Solde de recharge")) {
                        return@withContext executeConsumptionFetch(html)
                    }
                }
            } catch (_: Exception) {}
        }

        // Full login flow - Clear stale cookies first
        cookieJar.clear()

        // 1. Get login page for CSRF token
        val loginUrl = "https://espace-client.orange.ma/sso/login"
        val tokenHtml = redirectingClient.newCall(
            req(loginUrl).build()
        ).execute().body?.string() ?: ""
        
        var token = Jsoup.parse(tokenHtml).select("input[name=login_form[ezxform_token]]").`val`()

        // 2. POST Phone Number
        val post1Res = redirectingClient.newCall(
            req(loginUrl, loginUrl)
                .header("Origin", "https://espace-client.orange.ma")
                .post(FormBody.Builder()
                    .add("login_form[login]", loginId)
                    .add("login_form[ezxform_token]", token)
                    .build())
                .build()
        ).execute()
        
        val pwdHtml = post1Res.body?.string() ?: ""
        val token2 = Jsoup.parse(pwdHtml).select("input[name=login_form[ezxform_token]]").`val`()
        if (token2.isNotEmpty()) token = token2

        // 3. POST Password to /sso/check
        val checkUrl = "https://espace-client.orange.ma/sso/check"
        val checkRes = redirectingClient.newCall(
            req(checkUrl, post1Res.request.url.toString())
                .header("X-Requested-With", "XMLHttpRequest")
                .post(FormBody.Builder()
                    .add("login_form[_username]", loginId)
                    .add("login_form[_password]", pass)
                    .add("login_form[_remember_me]", "1")
                    .add("login_form[ezxform_token]", token)
                    .build())
                .build()
        ).execute()

        val checkBody = checkRes.body?.string() ?: ""
        
        if (checkRes.isSuccessful && checkBody.contains("\"status\":\"popup\"")) {
            val json = Gson().fromJson(checkBody, JsonObject::class.java)
            val popupHtml = json.get("message")?.asString ?: ""
            val popupDoc = Jsoup.parse(popupHtml)
            val ezxToken = popupDoc.select("input[name=select_msisdn_form[ezxform_token]]").`val`()
            val lines = popupDoc.select("input[name=select_msisdn_form[msisdn]]")
                .map { it.attr("value").trim() }
                .filter { it.isNotEmpty() }
                
            val cleanTarget = targetPhone.replace(" ", "")
            if (lines.contains(cleanTarget)) {
                return@withContext submitLineSelection(cleanTarget, ezxToken)
            } else if (lines.isNotEmpty()) {
                throw MultiLineException(lines, ezxToken, currentCookies)
            }
        } else if (!checkRes.isSuccessful || checkBody.contains("identifiants sont incorrects")) {
             throw IOException("Identifiant ou mot de passe incorrect")
        }

        // 4. Fetch consumption page
        val soldeHtml = redirectingClient.newCall(
            req("https://espace-client.orange.ma/mon-solde", "https://espace-client.orange.ma/").build()
        ).execute().body?.string() ?: ""

        executeConsumptionFetch(soldeHtml)
    }
}
